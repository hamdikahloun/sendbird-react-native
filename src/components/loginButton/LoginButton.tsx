import React, {memo} from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  Text,
  TouchableOpacity,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';

type LoginButtonProps = {
  onPress: () => void;
  disabled: boolean;
  loading: boolean;
};

const LoginButtonMemorized = ({
  onPress,
  disabled,
  loading,
}: LoginButtonProps) => {
  return (
    <LinearGradient style={styles.gradient} colors={['#6100FF', '#8940FF']}>
      <TouchableOpacity
        style={styles.button}
        testID="login-button"
        disabled={disabled}
        onPress={onPress}>
        {loading ? (
          <ActivityIndicator testID="loading-activityIndicator" color="white" />
        ) : (
          <Text testID="login-button-text" style={styles.buttonTex}>
            Login
          </Text>
        )}
      </TouchableOpacity>
    </LinearGradient>
  );
};

const styles = StyleSheet.create({
  gradient: {
    borderRadius: 10,
    height: 50,
    marginHorizontal: 16,
  },
  button: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
  },
  buttonTex: {
    color: 'white',
  },
});

/**
 * LoginButton - A Login Button component
 *
 * @example
 *
 * <LoginButton onPress={onPress} disabled={disabled} loading={loading} />
 */
export const LoginButton = memo(LoginButtonMemorized);
