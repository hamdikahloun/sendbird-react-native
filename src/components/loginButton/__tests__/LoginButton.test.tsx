import {render, fireEvent} from '@/../testing';
import React from 'react';
import {LoginButton} from '../LoginButton';

describe('LoginButton', () => {
  it('should render correctly', async () => {
    const onPress = jest.fn();
    const screen = render(
      <LoginButton onPress={onPress} disabled={false} loading={false} />,
    ).toJSON();
    expect(screen).toMatchSnapshot();
  });

  it('should call onPress if not disabled and not loading', async () => {
    const onPress = jest.fn();
    const {getByTestId} = render(
      <LoginButton onPress={onPress} disabled={false} loading={false} />,
    );
    fireEvent.press(getByTestId('login-button'));
    expect(onPress).toBeCalledTimes(1);
  });

  it('should not call onPress if disabled and not loading', async () => {
    const onPress = jest.fn();
    const {getByTestId} = render(
      <LoginButton onPress={onPress} disabled={true} loading={false} />,
    );
    fireEvent.press(getByTestId('login-button'));
    expect(onPress).toBeCalledTimes(0);
  });

  it('should show login text if not loading', async () => {
    const onPress = jest.fn();
    const {getByText} = render(
      <LoginButton onPress={onPress} disabled={false} loading={false} />,
    );
    expect(getByText('Login')).toBeTruthy();
  });

  it('should show loading state if loading', async () => {
    const onPress = jest.fn();
    const {findByTestId} = render(
      <LoginButton onPress={onPress} disabled={false} loading={true} />,
    );
    expect(findByTestId('loading-activityIndicator')).toBeTruthy();
  });
});
