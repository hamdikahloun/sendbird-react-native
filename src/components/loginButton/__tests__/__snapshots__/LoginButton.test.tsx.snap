// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginButton should render correctly 1`] = `
<BVLinearGradient
  colors={
    Array [
      "#6100FF",
      "#8940FF",
    ]
  }
  style={
    Object {
      "borderRadius": 10,
      "height": 50,
      "marginHorizontal": 16,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="login-button"
  >
    <Text
      style={
        Object {
          "color": "white",
        }
      }
      testID="login-button-text"
    >
      Login
    </Text>
  </View>
</BVLinearGradient>
`;
